// @gi-nx/gi-types/src/dag.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const filtering_1 = require("./filtering");
const nodes = [
    // specific start node (transform == undefined && inputHandles == undefined)
    {
        id: 'start_node',
        userDefinedCoordinates: [0, 500],
        outputHandles: [
            {
                id: 'primary'
            }
        ]
    },
    // specific end node (transform == undefined && outputHandles == undefined)
    {
        id: 'end_node',
        userDefinedCoordinates: [1500, 500],
        inputHandles: [
            {
                id: 'primary'
            }
        ]
    },
    {
        id: 'dag_carpark',
        transform: {
            type: 'parkingLots',
            parameters: {
                bayLength: 5,
                bayWidth: 5,
                roadWidth: 5,
                islandPeriod: 5,
                islandSize: 5,
                edgeOffset: 5,
                bayTwist: 5,
                longEdgeFlag: true
            }
        },
        userDefinedCoordinates: [1000, 500],
        inputHandles: [
            {
                id: 'boundary',
                label: 'Main Polygon',
                geoTypes: ['Polygon']
            },
            {
                id: 'footprints',
                label: 'Excl Polygon',
                geoTypes: ['Polygon']
            }
        ],
        outputHandles: [
            {
                id: 'primary',
                geoTypes: ['Polygon']
            }
        ]
    },
    {
        id: 'dag_filter1',
        transform: {
            type: 'filter',
            parameters: {
                filterDef: {
                    operator: filtering_1.AdvancedOperator.And,
                    filters: [
                        {
                            filter: {
                                type: 'string',
                                operator: filtering_1.SimpleOperator.Is,
                                value: 'car'
                            },
                            property: 'type'
                        }
                    ]
                }
            }
        },
        userDefinedCoordinates: [500, 0],
        inputHandles: [
            {
                id: 'primary',
                label: 'Features'
            }
        ],
        outputHandles: [
            {
                id: 'yes'
            },
            {
                id: 'no'
            }
        ]
    },
    {
        id: 'dag_filter_2',
        transform: {
            type: 'filter',
            parameters: {
                filterDef: {
                    operator: filtering_1.AdvancedOperator.And,
                    filters: [
                        {
                            filter: {
                                type: 'string',
                                operator: filtering_1.SimpleOperator.Is,
                                value: 'buildingSection'
                            },
                            property: 'type'
                        }
                    ]
                }
            }
        },
        userDefinedCoordinates: [500, 1000],
        inputHandles: [
            {
                id: 'primary',
                label: 'Features'
            }
        ],
        outputHandles: [
            {
                id: 'yes'
            },
            {
                id: 'no'
            }
        ]
    }
];
const edges = [
    {
        sourceNodeId: 'start_node',
        sourceHandleId: 'primary',
        destinationNodeId: 'dag_filter1',
        destinationHandleId: 'primary'
    },
    {
        sourceNodeId: 'start_node',
        sourceHandleId: 'primary',
        destinationNodeId: 'dag_filter2',
        destinationHandleId: 'primary'
    },
    {
        sourceNodeId: 'dag_filter1',
        sourceHandleId: 'yes',
        destinationNodeId: 'dag_carpark',
        destinationHandleId: 'boundary'
    },
    {
        sourceNodeId: 'dag_filter2',
        sourceHandleId: 'yes',
        destinationNodeId: 'dag_carpark',
        destinationHandleId: 'footprints'
    },
    {
        sourceNodeId: 'dag_carpark',
        sourceHandleId: 'primary',
        destinationNodeId: 'end_node',
        destinationHandleId: 'primary'
    }
];
