// @gi-nx/gi-types/src/filtering.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AdvancedOperator = exports.SimpleOperator = void 0;
var SimpleOperator;
(function (SimpleOperator) {
    SimpleOperator["Is"] = "is";
    SimpleOperator["IsNot"] = "is_not";
    SimpleOperator["Empty"] = "is_empty";
    SimpleOperator["NotEmpty"] = "is_not_empty";
    SimpleOperator["Contains"] = "contains";
    SimpleOperator["NotContains"] = "does_not_contain";
    SimpleOperator["StartsWith"] = "starts_with";
    SimpleOperator["EndsWith"] = "ends_with";
    SimpleOperator["GreaterThan"] = "greater_than";
    SimpleOperator["GreaterThanEq"] = "greater_than_or_equal";
    SimpleOperator["LessThan"] = "less_than";
    SimpleOperator["LessThanEq"] = "less_than_or_equal";
})(SimpleOperator || (exports.SimpleOperator = SimpleOperator = {}));
var AdvancedOperator;
(function (AdvancedOperator) {
    AdvancedOperator["And"] = "and";
    AdvancedOperator["Or"] = "or";
})(AdvancedOperator || (exports.AdvancedOperator = AdvancedOperator = {}));
