// @gi-nx/gi-types/src/layers.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.isMapboxTileErrorWithResponse = exports.GiraffeLayerType = exports.layerPurposes = void 0;
exports.layerPurposes = ['GIRDRW', 'GIRAPP', 'GIRPLA', 'GIRSPE', 'GIRMAR', 'GILENS', 'GIRTMP', 'GIRTER'];
// matches feasoapps/models
var GiraffeLayerType;
(function (GiraffeLayerType) {
    GiraffeLayerType[GiraffeLayerType["none"] = 0] = "none";
    // # vector layers,
    GiraffeLayerType[GiraffeLayerType["VectorUpload"] = 1] = "VectorUpload";
    GiraffeLayerType[GiraffeLayerType["FeatureServer"] = 2] = "FeatureServer";
    GiraffeLayerType[GiraffeLayerType["MapboxTiles"] = 3] = "MapboxTiles";
    // # 3d layers,
    GiraffeLayerType[GiraffeLayerType["CesiumTiles"] = 4] = "CesiumTiles";
    GiraffeLayerType[GiraffeLayerType["SceneServer"] = 5] = "SceneServer";
    // # raster layers, but may have legend and query endpoints,
    GiraffeLayerType[GiraffeLayerType["MapServer"] = 6] = "MapServer";
    GiraffeLayerType[GiraffeLayerType["WMS"] = 7] = "WMS";
    GiraffeLayerType[GiraffeLayerType["XYZ"] = 8] = "XYZ";
    GiraffeLayerType[GiraffeLayerType["Image"] = 9] = "Image";
    GiraffeLayerType[GiraffeLayerType["ImageServer"] = 10] = "ImageServer";
})(GiraffeLayerType || (exports.GiraffeLayerType = GiraffeLayerType = {}));
function isMapboxTileErrorWithResponse(err) {
    return !!err.status;
}
exports.isMapboxTileErrorWithResponse = isMapboxTileErrorWithResponse;
