// @gi-nx/gi-types/src/projectProperties.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.projectSystemSettings = void 0;
exports.projectSystemSettings = [
    {
        name: 'units',
        label: 'Units',
        defaultValue: 'meters',
        options: [
            { label: 'Meters', value: 'meters' },
            { label: 'Feet', value: 'feet' }
        ],
        type: 'select'
    },
    {
        name: 'currencySymbol',
        label: 'Currency symbol',
        defaultValue: '$',
        type: 'string'
    },
    {
        name: 'acronymGross',
        label: 'Gross area acronym',
        defaultValue: 'GBA',
        type: 'string'
    },
    {
        name: 'acronymNet',
        label: 'Net area acronym',
        type: 'string',
        defaultValue: 'GFA'
    },
    {
        name: 'acronymSaleable',
        label: 'Saleable area acronym',
        type: 'string',
        defaultValue: 'NSA'
    },
    {
        name: 'acronymFloorSpaceRatio',
        label: 'Floor space ratio acronym',
        type: 'string',
        defaultValue: 'FSR'
    },
    {
        name: 'mapboxStyle',
        defaultValue: 'mapbox://styles/mapbox/light-v10',
        description: 'Mapbox Basemap style (eg mapbox://styles/mapbox/light-v10)',
        label: 'Mapbox Basemap style',
        type: 'string'
    },
    {
        name: 'siteArea',
        label: 'Site Area',
        type: 'number',
        // defaultValue: '',
        dimension: 2,
        description: 'Manually override site area.',
        min: 0
    },
    {
        name: 'crs',
        defaultValue: '+proj=merc +a=6378137 +b=6378137 +lat_ts=0.0 +lon_0=0.0 +x_0=0.0 +y_0=0 +k=1.0 +units=m +nadgrids=@null +wktext  +no_defs',
        description: 'Coordinated Reference System for import and export in Proj4 syntax.',
        label: 'CRS',
        type: 'string'
    }
];
