// @gi-nx/gi-types/src/sectionProperties.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.isSectionWithPopup = exports.isStackedSectionWithMarker = exports.MarkerKind = void 0;
var MarkerKind;
(function (MarkerKind) {
    MarkerKind["Text"] = "text";
    MarkerKind["Icon"] = "icon";
    MarkerKind["Label"] = "label";
})(MarkerKind || (exports.MarkerKind = MarkerKind = {}));
function isStackedSectionWithMarker(obj) {
    var _a, _b;
    return (Object.values(MarkerKind).includes((_b = (_a = obj === null || obj === void 0 ? void 0 : obj.properties) === null || _a === void 0 ? void 0 : _a.marker) === null || _b === void 0 ? void 0 : _b.kind) ||
        // legacy
        obj.properties.markerHTML);
}
exports.isStackedSectionWithMarker = isStackedSectionWithMarker;
function isSectionWithPopup(obj) {
    var _a, _b;
    return (_b = (_a = obj === null || obj === void 0 ? void 0 : obj.properties) === null || _a === void 0 ? void 0 : _a.popup) === null || _b === void 0 ? void 0 : _b.content;
}
exports.isSectionWithPopup = isSectionWithPopup;
